### PR TITLE
Add XAttention reference operation

### PR DESCRIPTION
**Pull request by vshampor**
### Details:
 - Implements the XAttention operation using STL and OV reference operations.
 - Meant to be used for testing and debugging purposes only - does not bring performance benefits if used on its own. Actual HW-accelerated implementation would be using this code as reference.
 - #31955 contains an example of integrating with the CPU plugin.

### Tickets:
 - CVS-170365

### CODE REVIEW

File: src/core/reference/include/openvino/reference/xattention.hpp [hunk 1]
```diff
@@ -0,0 +1,349 @@
...
+                              reshaped_query_shape,
+                              reshaped_key_shape,
+                              out_shape,
+                              /* transpose_arg0 = */ false,
+                              /* transpose_arg1 = */ true);
+
+        size_t out_size = out_shape[0] * out_shape[1] * out_shape[2];
+
+        for (size_t i = 0; i < out_size; i++) {
+            out[i] = out[i] / (std::sqrt(reshaped_query_shape[2]) * m_stride);
```
Comment (l-bat) [inline src/core/reference/include/openvino/reference/xattention.hpp]: I didn't understand why it follows the formula: Q @ K^T / (sqrt(D) * S), since reshaped_query_shape[2] = head_size * stride

File: src/core/reference/include/openvino/reference/xattention.hpp [hunk 2]
```diff
@@ -0,0 +1,349 @@
...
+     * that the elements constitute a smallest subset constituting a sum portion no less than `threshold` of the total
+     * element sum.
+     * @param blocked_scores_data Pointer to the blocked score input.
+     * @param blocked_attention_scores_shape Shape of the blocked score input tensor. Expected shape is [num_heads,
+     * num_query_tokens / block_size, num_key_tokens / block_size]
+     * @return A vector of size `num_heads` of sets, each set containing pairs of block indices (.first is the block
+     * index along the query dimension, .second - along the key). Each set is the head-specific subset of blocks
+     * corresponding to the property described above.
+     */
+    XAttentionRetainedBlockIndicesForAllHeads get_block_indices_to_keep(const T* blocked_attention_scores_data,
```
Comment (l-bat) [inline src/core/reference/include/openvino/reference/xattention.hpp:216]: To make XAttention compatible with attention score-based eviction methods, we should enforce keeping last_query_size = 100 tokens, and calculate the attention mass excluding the last Q blocks to correctly apply the threshold for retained cumulative sum.

File: src/core/reference/include/openvino/reference/xattention.hpp [hunk 3]
```diff
@@ -0,0 +1,349 @@
...
+                    total_sum += current_score;
+                }
+            }
+            double cumsum = 0.0;
+            double required_sum = m_threshold * total_sum;
+            while (cumsum < required_sum && !indices_and_scores_queue.empty()) {
+                auto index_and_largest_score = indices_and_scores_queue.top();
+                indices_and_scores_queue.pop();
+                cumsum += index_and_largest_score.score;
+                total_sum += index_and_largest_score.score;
```
Comment (l-bat) [inline src/core/reference/include/openvino/reference/xattention.hpp]: Why do we need this line?

File: src/core/reference/include/openvino/reference/xattention.hpp [hunk 4]
```diff
@@ -0,0 +1,349 @@
...
+                retval[head_idx].insert(index_and_largest_score.idx);
+            }
+            while (!indices_and_scores_queue.empty()) {
+                auto index_and_largest_score = indices_and_scores_queue.top();
+                indices_and_scores_queue.pop();
+                total_sum += index_and_largest_score.score;
+            }
```
Comment (l-bat) [inline src/core/reference/include/openvino/reference/xattention.hpp]: I didn't get why we need this loop

File: src/core/reference/include/openvino/reference/xattention.hpp (hunk 3 above)
Comment (vshampor) [reply src/core/reference/include/openvino/reference/xattention.hpp]: Debug leftovers, removed

File: src/core/reference/include/openvino/reference/xattention.hpp (hunk 4 above)
Comment (vshampor) [reply src/core/reference/include/openvino/reference/xattention.hpp]: Debug leftovers, removed

File: src/core/reference/include/openvino/reference/xattention.hpp (hunk 2 above)
Comment (vshampor) [reply src/core/reference/include/openvino/reference/xattention.hpp:216]: I thought we disabled the xattention for a prompt chunk on the openvino.genai level if the chunk overlaps with the last dense token region:
https://github.com/openvinotoolkit/openvino.genai/blob/9ddbb07cfa48af645664280d095a1840e5c5bdcf/src/cpp/src/continuous_batching/scheduler.hpp#L613
In the PA op there is no information on what position each chunk occupies in the prompt, we'd have to set up an extra per-sequence input to the PA op just to tell xattention how many "last" tokens there are in the current chunk.

File: src/core/reference/include/openvino/reference/xattention.hpp (hunk 1 above)
Comment (vshampor) [reply src/core/reference/include/openvino/reference/xattention.hpp]: Fixed

File: src/core/tests/reference/xattention.cpp
Comment (praasz) [inline src/core/tests/reference/xattention.cpp:1]: The reference tests should be stored in:
`src/plugins/template/tests/functional/op_reference/`

as others.

File: src/core/reference/include/openvino/reference/xattention.hpp [hunk 5]
```diff
@@ -0,0 +1,349 @@
...
+                          const Shape& input_shape,
+                          T* output_data,
+                          const Shape& out_shape,
+                          bool is_antidiagonal) {
+        OPENVINO_ASSERT(input_shape.size() == 3);  // [num_heads, num_tokens, head_size]
+        OPENVINO_ASSERT(out_shape.size() == 3);
+        OPENVINO_ASSERT(input_shape[0] == out_shape[0]);
+        OPENVINO_ASSERT(input_shape[1] % m_stride == 0);
+        OPENVINO_ASSERT(input_shape[1] / m_stride == out_shape[1]);
+        OPENVINO_ASSERT(input_shape[2] * m_stride == out_shape[2]);
```
Comment (praasz) [inline src/core/reference/include/openvino/reference/xattention.hpp:77]: Remove these assert. the shapes and input validation is done during shape inference in operator implementation.

File: src/core/reference/include/openvino/reference/xattention.hpp [hunk 6]
```diff
@@ -0,0 +1,349 @@
...
+
+    /** Performs a softmax operation on the last dimension of the rank-3 input tensor.
+     * @param reshaped_qk_product_data Pointer to the reshaped query-key product input (attention logits pre-softmax).
+     * @param reshaped_qk_product_shape Shape of the input tensor. Expected shape is [num_heads, num_query_tokens /
+     * stride, num_key_tokens / stride].
+     * @param out Pointer to the output tensor data (attention scores)
+     * @param out_shape Shape of the output tensor data. Expected shape is strictly equal to
+     * `reshaped_qk_product_shape`.
+     */
+    void softmax(const T* reshaped_qk_product_data,
```
Comment (praasz) [inline src/core/reference/include/openvino/reference/xattention.hpp:153]: Why is required as it is just call for `ov::reference::softmax(reshaped_qk_product_data, out, reshaped_qk_product_shape, {2});`

File: src/core/tests/reference/xattention.cpp
Comment (vshampor) [reply src/core/tests/reference/xattention.cpp:1]: I am aware of these tests. All of these use utilize unnecessary abstractions of OV opset ops and involve the sample plugin to do unit tests. Seeing that there will never be a counterpart to this operation in the OV opset, there is no reason to proliferate the bad design decision to have all reference ops tested through the template plugin just for the sake of consistency.

File: src/core/reference/include/openvino/reference/xattention.hpp (hunk 5 above)
Comment (vshampor) [reply src/core/reference/include/openvino/reference/xattention.hpp:77]: Not relevant since there is no shape inference involved by design

File: src/core/reference/include/openvino/reference/xattention.hpp (hunk 6 above)
Comment (vshampor) [reply src/core/reference/include/openvino/reference/xattention.hpp:153]: This is a reference operation. It is designed to be understandable and testable first and foremost. I provide a set of functions that each directly map to a phase or sub-phase of the intended HW-accelerated kernel. Softmax is one of these phases and therefore has a separate function. It also has an extra shape check, so having a function wrapper has some utility.

File: src/core/reference/include/openvino/reference/xattention.hpp [hunk 7]
```diff
@@ -0,0 +1,346 @@
...
+                    size_t key_block_idx = key_len_idx / antidiagonals_per_xattention_block;
+                    auto target_block_sum_ptr = out + out_head_offset + query_block_idx * out_shape[2] + key_block_idx;
+                    *target_block_sum_ptr += *(attention_scores_data + in_head_offset +
+                                               query_len_idx * attention_scores_shape[2] + key_len_idx);
+                }
+            }
+        }
+    }
+
+    /** Selects the elements of the input tensor along the last two dimensions, indepently along the first dimension, so
```
Comment (l-bat) [inline src/core/reference/include/openvino/reference/xattention.hpp]:
```suggestion
    /** Selects the elements of the input tensor along the last two dimensions, independently along the first dimension, so
```

Comment (vshampor) [reply src/core/reference/include/openvino/reference/xattention.hpp]: Done

File: src/core/reference/include/openvino/reference/xattention.hpp [hunk 8]
```diff
@@ -0,0 +1,346 @@
...
+            }
+            double cumsum = 0.0;
+            double required_sum = m_threshold * total_sum;
+            while (cumsum < required_sum && !indices_and_scores_queue.empty()) {
+                auto index_and_largest_score = indices_and_scores_queue.top();
+                indices_and_scores_queue.pop();
+                cumsum += index_and_largest_score.score;
+                retval[head_idx].insert(index_and_largest_score.idx);
+            }
+        }
```
Comment (WeldonWangwang) [inline src/core/reference/include/openvino/reference/xattention.hpp:251]: Hi @vshampor It looks like this implementation selects blocks by globally sorting all blocks per head and accumulating their scores until the threshold is reached. In the X-Attention paper (Section 2.2, [https://arxiv.org/abs/2503.16428](https://arxiv.org/abs/2503.16428?utm_source=chatgpt.com)
), the algorithm selects blocks per query block based on antidiagonal scoring. Could this global accumulation approach lead to differences from the paper’s intended block coverage for each query(https://github.com/mit-han-lab/x-attention/blob/e37988770b9d1bebd489eba011d615f35587ba08/xattn/src/utils.py#L44)?

Comment (vshampor) [reply src/core/reference/include/openvino/reference/xattention.hpp:251]: Please consider the relative position of this function in the frontend `select_blocks` function and observe that the block selection is ultimately done among the antidiagonal subset of the full attention matrix. I didn't understand the concerns about globality - in which sense do you consider the sorting to be global and why does it not match to the original algo?

Comment (WeldonWangwang) [reply src/core/reference/include/openvino/reference/xattention.hpp:251]: Simplify the problem, I use[ https://github.com/mit-han-lab/x-attention/tree/main?tab=readme -Use your test case data in ov file # quick use ](https://github.com/mit-han-lab/x-attention/tree/main?tab=readme-ov-file#quick-use)  with your test date in https://github.com/openvinotoolkit/openvino/blob/231ae69602b6cc25e8da34adf692dade9941cdda/src/core/tests/reference/xattention.cpp#L27 The selected blocks are inconsistent:
```
THRESHOLD = 0.8
BLOCK_SIZE = 2
STRIDE = 2

# batch=1, head=2, seq_len=4, dim=4
bsz = 1
heads = 2
seq_len = 4
dim = 4

q_data = [
     3.144,  8.512,  8.518, -8.386,
     7.889, -5.721,  5.507,  4.295,
    -6.624, -8.463,  7.474,  9.879,
     4.534, -5.908, -9.388,  2.356,

     7.497,  8.186, -8.658, -4.796,
    -8.248, -9.797, -7.907, -4.513,
     3.469,  7.633,  7.244, -6.844,
    -7.173,  4.450,  6.705, -7.035
]

q = torch.tensor(q_data, dtype=torch.bfloat16).reshape(bsz, heads, seq_len, dim).to("cuda")
k = q.clone()  # K = Q
v = torch.randn((bsz, heads, seq_len, dim), dtype=torch.bfloat16).to("cuda")

attention_output = Xattention_prefill(
    query_states=q,
    key_states=k,
    value_states=v,
    stride=STRIDE,
    block_size=BLOCK_SIZE,
    use_triton=True,
    chunk_size=4
)
```

your code:
Head 0 selected blocks: (0,1) (1,0)
Head 1 selected blocks: (0,1) (1,0)
mit-han-lab/x-attention:
Head 0 selected blocks: (0,0) (1,0) (1,1)
Head 1 selected blocks: (0,0) (1,0) (1,1)
So may I know what is the source of the reference results for your test?
